### ISSUE TITLE
Android 4.4 ZimWiki detection crash when opening .txt file

### ISSUE THREAD

**Issue by AnniePerson** [gsantner/markor#1328]
#### General information

* **App version: 2.6.0 Google PlayStore** 
* **System:Android 5.0 vs. 4.4** 

#### Description
I cannot open .txt files on my tablet. Only a message '"Markor" wurde beendet' is displayed and Markor is closed. This happens with every .txt file on my tablet with Android 4.4. On my phone with Android 5.0 Markor works fine. 
No difference if I try to open a .txt from Markor's "Files" section or by applying "open with" on a .txt file from an Android file parser. The same file renamed to .md is opened correctly. 
I activated "Experimental features" and "Debug" in Markor settings but could not find a debug log.

**Comment by harshad1** [comment]
@gsantner I believe I can reproduce this on the emulator:

```
E/AndroidRuntime: FATAL EXCEPTION: main
    Process: net.gsantner.markor_test, PID: 2293
    java.lang.ExceptionInInitializerError
        at net.gsantner.markor.format.zimwiki.ZimWikiTextConverter.isFileOutOfThisFormat(ZimWikiTextConverter.java:184)
        at net.gsantner.markor.util.DocumentIO.loadDocument(DocumentIO.java:130)
        at net.gsantner.markor.activity.DocumentEditFragment.loadDocument(DocumentEditFragment.java:556)
        at net.gsantner.markor.activity.DocumentEditFragment.onViewCreated(DocumentEditFragment.java:182)
        at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1471)
        at android.support.v4.app.FragmentManagerImpl.moveFragmentToExpectedState(FragmentManager.java:1784)
        at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1852)
        at android.support.v4.app.BackStackRecord.executeOps(BackStackRecord.java:802)
        at android.support.v4.app.FragmentManagerImpl.executeOps(FragmentManager.java:2625)
        at android.support.v4.app.FragmentManagerImpl.executeOpsTogether(FragmentManager.java:2411)
        at android.support.v4.app.FragmentManagerImpl.removeRedundantOperationsAndExecute(FragmentManager.java:2366)
        at android.support.v4.app.FragmentManagerImpl.execPendingActions(FragmentManager.java:2273)
        at android.support.v4.app.FragmentManagerImpl.dispatchStateChange(FragmentManager.java:3273)
        at android.support.v4.app.FragmentManagerImpl.dispatchActivityCreated(FragmentManager.java:3229)
        at android.support.v4.app.FragmentController.dispatchActivityCreated(FragmentController.java:201)
        at android.support.v4.app.FragmentActivity.onStart(FragmentActivity.java:620)
        at android.support.v7.app.AppCompatActivity.onStart(AppCompatActivity.java:178)
        at android.app.Instrumentation.callActivityOnStart(Instrumentation.java:1171)
        at android.app.Activity.performStart(Activity.java:5241)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2168)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2245)
        at android.app.ActivityThread.access$800(ActivityThread.java:135)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1196)
        at android.os.Handler.dispatchMessage(Handler.java:102)
        at android.os.Looper.loop(Looper.java:136)
        at android.app.ActivityThread.main(ActivityThread.java:5017)
        at java.lang.reflect.Method.invokeNative(Native Method)
        at java.lang.reflect.Method.invoke(Method.java:515)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:779)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:595)
        at dalvik.system.NativeStart.main(Native Method)
     Caused by: java.util.regex.PatternSyntaxException: Look-behind pattern matches must have a bounded maximum length near index 13:
    (?<=((
    |^)\s{0,16}))\*(?= )
                 ^
```

**Comment by whunger** [comment]
I get the same exception on my Android 4.4.4 device:

```
05-03 11:21:23.442 W/dalvikvm(28867): Exception Ljava/util/regex/PatternSyntaxException; thrown while initializing Lnet/gsantner/markor/format/zimwiki/ZimWikiHighlighter$Patterns;
05-03 11:21:23.462 D/AndroidRuntime(28867): Shutting down VM
05-03 11:21:23.462 W/dalvikvm(28867): threadid=1: thread exiting with uncaught exception (group=0x41d63ba8)
05-03 11:21:23.462 E/AndroidRuntime(28867): FATAL EXCEPTION: main
05-03 11:21:23.462 E/AndroidRuntime(28867): Process: net.gsantner.markor, PID: 28867
05-03 11:21:23.462 E/AndroidRuntime(28867): java.lang.ExceptionInInitializerError
05-03 11:21:23.462 E/AndroidRuntime(28867): 	at net.gsantner.markor.format.zimwiki.ZimWikiTextConverter.isFileOutOfThisFormat(ZimWikiTextConverter.java:184)
05-03 11:21:23.462 E/AndroidRuntime(28867): 	at net.gsantner.markor.format.TextFormat.isTextFile(TextFormat.java:62)
05-03 11:21:23.462 E/AndroidRuntime(28867): 	at net.gsantner.markor.activity.MainActivity$1.onFsViewerSelected(MainActivity.java:413)
05-03 11:21:23.462 E/AndroidRuntime(28867): 	at net.gsantner.opoc.ui.FilesystemViewerFragment.onFsViewerSelected(FilesystemViewerFragment.java:186)
05-03 11:21:23.462 E/AndroidRuntime(28867): 	at net.gsantner.opoc.ui.FilesystemViewerAdapter.onClick(FilesystemViewerAdapter.java:337)
05-03 11:21:23.462 E/AndroidRuntime(28867): 	at android.view.View.performClick(View.java:4449)
05-03 11:21:23.462 E/AndroidRuntime(28867): 	at android.view.View$PerformClick.run(View.java:18631)
05-03 11:21:23.462 E/AndroidRuntime(28867): 	at android.os.Handler.handleCallback(Handler.java:733)
05-03 11:21:23.462 E/AndroidRuntime(28867): 	at android.os.Handler.dispatchMessage(Handler.java:95)
05-03 11:21:23.462 E/AndroidRuntime(28867): 	at android.os.Looper.loop(Looper.java:136)
05-03 11:21:23.462 E/AndroidRuntime(28867): 	at android.app.ActivityThread.main(ActivityThread.java:5001)
05-03 11:21:23.462 E/AndroidRuntime(28867): 	at java.lang.reflect.Method.invokeNative(Native Method)
05-03 11:21:23.462 E/AndroidRuntime(28867): 	at java.lang.reflect.Method.invoke(Method.java:515)
05-03 11:21:23.462 E/AndroidRuntime(28867): 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:785)
05-03 11:21:23.462 E/AndroidRuntime(28867): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:601)
05-03 11:21:23.462 E/AndroidRuntime(28867): 	at dalvik.system.NativeStart.main(Native Method)
05-03 11:21:23.462 E/AndroidRuntime(28867): Caused by: java.util.regex.PatternSyntaxException: Look-behind pattern matches must have a bounded maximum length near index 13:
05-03 11:21:23.462 E/AndroidRuntime(28867): (?<=((
05-03 11:21:23.462 E/AndroidRuntime(28867): |^)\s{0,16}))\*(?= )
05-03 11:21:23.462 E/AndroidRuntime(28867):              ^
05-03 11:21:23.462 E/AndroidRuntime(28867): 	at java.util.regex.Pattern.compileImpl(Native Method)
05-03 11:21:23.462 E/AndroidRuntime(28867): 	at java.util.regex.Pattern.compile(Pattern.java:411)
05-03 11:21:23.462 E/AndroidRuntime(28867): 	at java.util.regex.Pattern.<init>(Pattern.java:394)
05-03 11:21:23.462 E/AndroidRuntime(28867): 	at java.util.regex.Pattern.compile(Pattern.java:381)
05-03 11:21:23.462 E/AndroidRuntime(28867): 	at net.gsantner.markor.format.zimwiki.ZimWikiHighlighter$Patterns.<clinit>(ZimWikiHighlighter.java:38)
05-03 11:21:23.462 E/AndroidRuntime(28867): 	... 16 more
05-03 11:21:23.502 W/ActivityManager( 2767):   Force finishing activity net.gsantner.markor/.activity.MainActivity
```

It would be great if this could be fixed! Markor would be one of the few really useful gems running on this device.

**Comment by harshad1** [comment]
@fredericjacob I believe you may understand what is going on here

**Comment by fredericjacob** [comment]
It seems that the regex engine works different in Android 4.4 than in the later versions. So the issue is that the lookbehind pattern `(?<=((\n|^)\s{0,16}))` contains a variable quantifier `{0,16}`, which is supported in later versions of Android, but not in 4.4 (and below, probably). I'm going to take a closer look at this in the weekend and work on a solution.

**Comment by fredericjacob** [comment]
I created a pull request that fixes this issue, see #1341.
